# HW 2

## Instructions

* The course will be using an automatic grading system.
* After each question there will appear a code block with some prepared code to add your answer to a dictionary that will be converted into a data frame, saved as CSV and uploaded to moodle for grading.
* Please do not edit any code other than in placeholders marked `#### your code here ####`
* Don't forget to run the code block after you write your answer.
* You can add code blocks wherever you want in order to interact with datasets and play with your own code.
* In the next code block please fill in your ID number and email account as strings in the appropriate places.
* And don't forget to run the block!

In [133]:
ans = {}
ans['HW'] = 'HW2'
ans['id_number'] = '318903028'
ans['email'] = 'matantalvi@mail.tau.ac.il'

## Avocado Data

For this assignment you will need to import two csv files which can be found in moodle: `avocade_price.csv` which contains 2015-2018 daily price for avocado across the US, and `avocado_bags.csv` which contains 2015-2018 daily no. of avocado bags sold across the US.

## Pandas

* In this section we will be interacting with a real dataset.
* This next block loads the dataset from the repository.
* Remember you are able to add code blocks to interact with the dataset.
* If at any point you think that you might have altered the original data, you can simply rerun this code block to load a fresh copy.
* Please do not change the blocks that record your answer to the `ans` dictionary.
* The pandas documentation, and specifically the list of `DataFrame` methods might be helpful, [here](https://pandas.pydata.org/pandas-docs/stable/reference/frame.html).

In [134]:
import pandas as pd

# This assumes you have avocado_fixed.csv in the notebook folder
# If you're working with a folder in Gooogle Drive and want to read from it - 
# Mount the drive, either by pressing the mount drive button or by running:
# from google.colab import drive
# drive.mount('/content/drive')
# then, reading the csv from the relevant `path` with
# df_avocado = pd.read_csv(path + 'avocado_fixed.csv')

df_avocado = pd.read_csv('avocado_price.csv')
print('Avocado prices data has %d rows and %d columns' % df_avocado.shape)
df_avocado.head()

Avocado prices data has 9126 rows and 4 columns


Date  price  year  region
0  12/27/15   1.33  2015  Albany
1  12/20/15   1.35  2015  Albany
2  12/13/15   0.93  2015  Albany
3   12/6/15   1.08  2015  Albany
4  11/29/15   1.28  2015  Albany

### Q1.
What is the median of the `year` column?

1. 2015

2. 2016

3. 2017

4. 2018

In [135]:
x = df_avocado['year'].median()
x

2016.0

In [136]:
ans['Q1'] = 2

### Q2.
On what date and region was the price the highest?

1. PhoenixTucson - 2/5/17

2. Chicago - 10/8/17

3. SanFrancisco - 10/30/16

4. LosAngeles 2/5/17

__Hint:__ the pandas data frame method `sort_values()` can come in handy, look it up in the pandas documentation.

In [137]:
max_price = df_avocado['price'].max()
df_avocado.sort_values('price')

Date  price  year         region
7412    2/5/17   0.46  2017  PhoenixTucson
1716  12/27/15   0.49  2015  PhoenixTucson
6622    1/1/17   0.51  2017        Houston
1752   4/19/15   0.51  2015  PhoenixTucson
4566    3/6/16   0.51  2016  PhoenixTucson
...        ...    ...   ...            ...
6053   10/1/17   2.14  2017        Chicago
6055   9/17/17   2.15  2017        Chicago
6370   10/8/17   2.19  2017    GrandRapids
5000  10/30/16   2.20  2016   SanFrancisco
6052   10/8/17   2.22  2017        Chicago

[9126 rows x 4 columns]

In [138]:
ans['Q2'] = 2

### Q3.
How many days are there in the data for the `SanFrancisco` region, in the year 2018?

1. 5
2. 12
3. 13
4. 53

In [139]:
#forestfires.loc[(forestfires['month'].isin(['mar', 'oct'])) & (forestfires['X'] > 7)]
df_avocado.loc[(df_avocado['region'] == 'SanFrancisco') & (df_avocado['year'] == 2018)].count()

Date      12
price     12
year      12
region    12
dtype: int64

In [140]:
ans['Q3'] = 2

### Q4.
Which region had the lowest price for the entire period? (on average)

1. Albany

2. HartfordSpringfield

3. Houston

4. PhoenixTucson

__Hint:__ `groupby` and `agg`.

In [141]:
df_avocado.groupby('region').agg({'price':'mean'}).sort_values('price')

price
region                       
PhoenixTucson        0.728225
Houston              0.825089
WestTexNewMexico     0.842130
DallasFtWorth        0.846450
SouthCentral         0.869408
LosAngeles           0.976450
West                 0.985089
Nashville            1.012899
LasVegas             1.012959
CincinnatiDayton     1.015503
Portland             1.046509
SanDiego             1.061479
Columbus             1.066272
Atlanta              1.068817
Denver               1.073964
Boise                1.076036
NewOrleansMobile     1.085266
TotalUS              1.092012
Roanoke              1.096036
Louisville           1.105325
California           1.105444
Spokane              1.115976
RichmondNorfolk      1.119822
Detroit              1.123964
Indianapolis         1.144852
SouthCarolina        1.146213
Southeast            1.163018
Plains               1.165503
Seattle              1.169763
GreatLakes           1.181893
StLouis              1.185740
Jacksonville         1.193609
Tampa                1.201598
Midsouth             1.207160
Orlando              1.214438
RaleighGreensboro    1.227101
Pittsburgh           1.250651
MiamiFtLauderdale    1.254320
HarrisburgScranton   1.258817
NorthernNewEngland   1.260237
Sacramento           1.273964
Charlotte            1.275089
Boston               1.304379
GrandRapids          1.325030
BaltimoreWashington  1.344201
Northeast            1.344438
Albany               1.348757
Chicago              1.369349
BuffaloRochester     1.382308
Syracuse             1.386923
Philadelphia         1.396331
SanFrancisco         1.397160
NewYork              1.402130
HartfordSpringfield  1.408047

In [142]:
ans['Q4'] = 4

### Q5.
In what year were avocado prices the lowest? (on average)

1. 2015

2. 2016

3. 2017

4. 2018

In [143]:
df_avocado.groupby('year').agg({'price':'mean'}).sort_values('price')

price
year          
2015  1.077963
2016  1.105595
2018  1.127886
2017  1.294888

In [144]:
ans['Q5'] = 1

### Q6.
In what year were the prices lowest in Indianapolis? (on average)

1. 2015

2. 2016

3. 2017

4. 2018

In [145]:
df_avocado.loc[(df_avocado['region'] == 'Indianapolis')].groupby('year').agg({'price':'mean'}).sort_values('price')

price
year          
2018  0.985000
2015  1.104615
2016  1.118077
2017  1.246792

In [146]:
ans['Q6'] = 4

### Q7.
What is the median price in the Columbus region in 2017?

1. 1.015

2. 1.030

3. 1.100

4. 1.163

In [147]:
df_avocado.loc[(df_avocado['region'] == 'Columbus') & (df_avocado['year'] == 2017)]['price'].median()

1.1

In [148]:
ans['Q7'] = 3

---

We will now load some additional data to play with. The new data set holds number of avocado bags sold in each region for the same period.

In [149]:
### Do whatever you have to do to get that file from Moodle ###

df_avocado_bags = pd.read_csv('avocado_bags.csv')
print('Avocado bags data has %d rows and %d columns' % df_avocado_bags.shape)
df_avocado_bags.head()

Avocado bags data has 9126 rows and 4 columns


Date  num_bags  year  region
0  12/27/15   8696.87  2015  Albany
1  12/20/15   9505.56  2015  Albany
2  12/13/15   8145.35  2015  Albany
3   12/6/15   5811.16  2015  Albany
4  11/29/15   6183.95  2015  Albany

__Notice:__
* To answer the following questions you will need to join the two datasets together properly and make calculations on columns from both `DataFrame`s.
* Assume "avocado market" for a single date can be measured by price times number of bags (not a trick question).
* Note each region has only one observation on each date, and we don't have missing values - so there are no weird/spooky/shady things hiding in the data. Yet.

### Q8.
Which **of the following** regions has a larger avocado market? (in total, over the entire period)

1. California

2. NewYork

3. Chicago

4. LosAngeles

In [150]:
df_market = pd.merge(df_avocado,df_avocado_bags, how='inner', on=['region', 'Date'])
df_market['avocado market'] = df_market['num_bags'] * df_market['price']
assist = df_market.loc[(df_market['region'].isin(['California','NewYork','Chicago','LosAngeles']))]
assist.groupby('region').agg({'avocado market': 'sum'}).sort_values('avocado market').head()

avocado market
region                    
Chicago       2.192339e+07
NewYork       9.107116e+07
LosAngeles    1.503435e+08
California    2.554244e+08

In [151]:
ans['Q8'] = 1

### Q9.
What year were avocado sales the highest? (in total, by avocado market, ignore that maybe data is missing for some years)

1. 2015

2. 2016

3. 2017

4. 2018

In [152]:
assist = df_market.loc[(df_market['year_x'].isin([2015,2016,2017,2018]))]
assist.groupby('year_x').agg({'avocado market': 'sum'}).sort_values('avocado market').head()

avocado market
year_x                
2018      4.812983e+08
2015      7.730535e+08
2016      1.437407e+09
2017      1.822195e+09

In [153]:
ans['Q9'] = 3

### Q10.
How many regions have a total avocado market above `15e06` in 2016?

1. 1
2. 11
3. 14
4. 26

In [154]:
x = 15*(10**6)
assist = df_market.loc[(df_market['year_x'] == 2016)]
assist= assist.groupby('region').agg({'avocado market': 'sum'}).sort_values('avocado market')
assist.loc[(assist['avocado market'] > x)].shape


(14, 1)

In [155]:
ans['Q10'] = 3

### Q11.

All regions have the same number of days reported in the year 2017 (challenge yourself to do this with as little code as possible)

1. True
2. False

In [156]:
df_market.loc[(df_market['year_x'] == 2017)].groupby('region').agg({'year_x':'sum'})['year_x'].unique().size


1

In [157]:
ans['Q11'] = 1

## Data Science Basics

The following questions reference material from the first lecture and should be answered after the lecture.

### Q12.
Which of the following is not primarily a data science project: 

1.	Collect and analyze data from a linear accelerator, looking for a new particle

2.	Build a robot which serves cocktail drinks

3.	Given two website designs, test which one attracts more traffic

4.	Do an online poll to predict the outcome of an election


In [158]:
ans['Q12'] = 2

We mentioned in class that data science has three interweaved components:
* Computing expertise
* Statistical expertise
* Domain expertise (e.g. Physics)

Consider the following tasks involved in a hypothetical data science project to find a new particle: 

1.	Designing the database to hold results

2.	Deciding which signals should be collected, and what kind of signals we are looking for to identify a new particle

3.	Fitting regression models to the data and performing statistical tests in search of the signals

4.  Recruiting the right personnel to carry out the project 

### Q13.
Which one of the above most clearly belongs to the statistical expertise category?
### Q14.
Which one most clearly belongs to the computing expertise category?
### Q15.
Which one most clearly belongs to the domain expertise category?

In [159]:
ans['Q13'] = 3
ans['Q14'] = 1
ans['Q15'] = 2

# Finish!

* To submit your HW please run this last code block and follow the instructions.
* This code will create a CSV file in the current directory on your local machine or in your google drive (Google Colab)
* Please download it and upload it to the moodle HW1
* Make sure that you write your ID correctly!

In [160]:
import pandas as pd

df_ans = pd.DataFrame.from_dict(ans, orient='index')
if df_ans.shape[0] == 18:
  df_ans.to_csv('{}_{}.csv'.format(ans['HW'],str(ans['id_number'])))
else:
    print("seems like you missed a question, make sure you have run all the code blocks")

**For Google Colab users who want to save their CSV in their drive**

* Run the following script.
* Note that you will need to authorize GD once. For more information: https://colab.research.google.com/notebooks/io.ipynb (under Mounting Google Drive locally)
* You may change the path to be in your inner folder at GD, or not use GD at all and just manually download the CSV file to your local machine.
* Eventually upload the CSV to Moodle.

In [161]:
from google.colab import drive
drive.mount('/content/drive')

# Three options of the path of the file 
path ='/content/drive/MyDrive/' # general outer folder
# path = '/content/drive/MyDrive/Colab Notebooks/Intro_DS_2023/HW/HW2/' # assuming an inner folder called Intro_DS_2023 etc.
# path = './' # here, in this notebook environment

import pandas as pd 
df_ans = pd.DataFrame.from_dict(ans, orient='index')

if df_ans.shape[0] == 18:
  df_ans.to_csv(path+'{}_{}.csv'.format(ans['HW'],str(ans['id_number'])))
else:
    print("seems like you missed a question, make sure you have run all the code blocks")


ModuleNotFoundError: No module named 'google'